virtual_rainforest/models/abiotic_simple/simple_regression.py

```diff
@@ -0,0 +1,571 @@
+    """
+
+    # TODO: make sure variables are representing correct time interval, see #222
+    output = {}
+
+    # Initialise DataArrays filled with NaN
+    output["air_temperature"] = DataArray(
+        np.full((len(layer_roles), len(data.grid.cell_id)), np.nan),
+        dims=["layers", "cell_id"],
+        coords={
+            "layers": np.arange(0, len(layer_roles)),
+            "layer_roles": ("layers", layer_roles),
+            "cell_id": data.grid.cell_id,
+        },
+        name="air_temperature",
+    )
+
+    # copy and assign new variable names
+    for name in (
+        "relative_humidity",
+        "vapour_pressure_deficit",
+        "soil_temperature",
+        "atmospheric_pressure",
+        "atmospheric_co2",
+        "surface_runoff",
+    ):
+        output[name] = output["air_temperature"].copy().rename(name)
+
+    # The initial soil moisture constant in all soil layers, all other layers set to NaN
+    soil_moisture = output["air_temperature"].copy().rename("soil_moisture")
+    output["soil_moisture"] = soil_moisture.where(
+        soil_moisture.coords["layer_roles"] != "soil",
+        initial_soil_moisture,
+    )
+
+    # TODO until the plant model is ready, these variables are initialised here
+    layer_heights = xr.concat(
+        [
+            DataArray([[32, 32, 32], [30, 30, 30], [20, 20, 20], [10, 10, 10]]),
+            DataArray(np.full((7, 3), np.nan)),
+            DataArray([[1.5, 1.5, 1.5], [0.1, 0.1, 0.1]]),
+            DataArray([[-0.1, -0.1, -0.1], [-1, -1, -1]]),
+        ],
+        dim="dim_0",
+    )
+    output["layer_heights"] = layer_heights.rename(
+        {"dim_0": "layers", "dim_1": "cell_id"}
+    ).assign_coords(
+        {
+            "layers": np.arange(0, 15),
+            "layer_roles": ("layers", layer_roles),
+            "cell_id": data.grid.cell_id,
+        }
+    )
+
+    leaf_area_index = xr.concat(
+        [
+            DataArray(np.full((1, 3), np.nan)),
+            DataArray(np.full((1, 3), 3)),
+            DataArray(np.full((13, 3), np.nan)),
+        ],
+        dim="dim_0",
+    )
+    output["leaf_area_index"] = leaf_area_index.rename(
+        {"dim_0": "layers", "dim_1": "cell_id"}
+    ).assign_coords(
+        {
+            "layers": np.arange(0, 15),
+            "layer_roles": ("layers", layer_roles),
+            "cell_id": data.grid.cell_id,
+        }
+    )
+
+    return output
+
+
+def run_simple_regression(
+    data: Data,
+    layer_roles: List[str],
+    time_index: int,  # could be datetime?
+    MicroclimateGradients: Dict[str, float] = MicroclimateGradients,
+    Bounds: Dict[str, float] = Bounds,
+    water_interception_factor: Union[DataArray, float] = MicroclimateParameters[
+        "water_interception_factor"
+    ],
+    soil_moisture_capacity: Union[DataArray, float] = MicroclimateParameters[
+        "soil_moisture_capacity"
+    ],
+) -> Dict[str, DataArray]:
+    r"""Calculate simple microclimate.
```

Comment:
This function feels like it is the content of `update` but outsourced into a stand alone function. I think it is probably better if the main sequence stays in the model, possibly in extra methods, and external static functions are only used for smaller steps.

Reply:
Yeah, maybe. I'm never sure with this things, to be honest. I tend to think of methods as relatively small functions that either update the state of the object or provide some output to be used elsewhere that make use of the object state. This function do the first thing, but is pretty large, so it does not fit into that definition (a personal definition). When in doubt, I tend to choose the option that makes testing easier.

Reply:
Could you please explain this a bit more? I was trying to keep the `abiotic_simple_model` as tidy as possible and only call very few functions that do all the bits. Do you mean that I should leave only the 'helper functions' in the `simple_regression` and move all the content of `run_simple_regression` to the `AbioticSimpleModel.update()`?

Reply:
This is _definitely_ a personal choice. 

My feeling is that `run_simple_regression` does far more than running a single simple regression and that this function is basically the `update` method, but operating a bit indirectly. It has to pass back a list of updated outputs and then copy them into the object, rather being able to access `self.data` directly. 

We don't want big complex methods, but my comment was motivated by wanting to be able to read the `update` method to get a high level view of the steps being taken. The `run_simple_regression` function  runs a relatively small number of calls to existing low level functions that are easy to test, so packing them all up in this function and then feeding that into `update` felt like an extra layer. The `update` method would not be huge if it just worked through this sequence, directly calling the low level functions and assigning the results into `self.data`.

That's where I was coming from, but as @dalonsoa says, this is very much a personal style thing.

Reply:
Your comment makes a lot of sense. I think I could have chosen a better name for the function, it is more like `run_simple_model` which is, like you say, basically `update()`. I think I chose this approach based on the process-based model where I had multiple files that contained functions for one type of process, which is not necessary here. I'll have a think about it and do the other parts first :-)

Reply:
I keep this as it is for now because there will likely be changes to the structure and for now, I find it easier to navigate. I added a comment in the `abiotic_simple_model `that this routine could be unnested and moved to `update()`